nameserver/addrs.go

```diff
@@ -0,0 +1,99 @@
+package nameserver
+
+import (
+	"bytes"
+	"fmt"
+	. "github.com/weaveworks/weave/common"
+	"net"
+)
+
+// A zone record
+// Note: we will try to keep all the zone records information is concentrated here. Maybe not all queries will
+//       use things like priorities or weights, but we do not want to create a hierarchy for dealing with all
+//       possible queries...
+type ZoneRecord interface {
+	Name() string  // The name for this IP
+	IP() net.IP    // The IP (v4) address
+	Priority() int // The priority
+	Weight() int   // The weight
+}
+
+type ZoneLookup interface {
+	// Lookup for a name
+	LookupName(name string) ([]ZoneRecord, error)
+	// Lookup for an address
+	LookupInaddr(inaddr string) ([]ZoneRecord, error)
+}
+
+type ZoneObserver interface {
+	// Observe anything that affects a particular name in the zone
+	ObserveName(name string, observer ZoneRecordObserver) error
+	// Observe anything that affects a particular IP in the zone
+	ObserveInaddr(inaddr string, observer ZoneRecordObserver) error
+}
+
+type ZoneRecordObserver func(ZoneRecord)
+
+/////////////////////////////////////////////////////////////
+
+// A basic record struct (satisfying the ZoneRecord interface)
+type Record struct {
+	name     string
+	ip       net.IP
+	priority int
+	weight   int
+}
+
+func (r Record) Name() string  { return r.name }
+func (r Record) IP() net.IP    { return r.ip }
+func (r Record) Priority() int { return r.priority }
+func (r Record) Weight() int   { return r.weight }
+
+func (i Record) String() string {
+	var buf bytes.Buffer
+	if len(i.Name()) > 0 {
+		buf.WriteString(fmt.Sprintf("%s", i.Name()))
+	}
+	if !i.IP().IsUnspecified() {
+		buf.WriteString(fmt.Sprintf("[%s]", i.IP()))
+	}
+	if i.Priority() > 0 {
+		buf.WriteString(fmt.Sprintf("/P:%d", i.Priority()))
+	}
+	if i.Weight() > 0 {
+		buf.WriteString(fmt.Sprintf("/W:%d", i.Weight()))
+	}
+
+	return buf.String()
+}
+
+/////////////////////////////////////////////////////////////
+
+// simple IPv4 type that can be used as a key in a map (in contrast with net.IP), used for sets of IPs, etc...
+type IPv4 [4]byte
+
+func (ip IPv4) toNetIP() net.IP { return net.IPv4(ip[0], ip[1], ip[2], ip[3]) }
+func (ip IPv4) String() string  { return ip.toNetIP().String() }
+func ipToIPv4(nip net.IP) IPv4  { ip := nip.To4(); return IPv4([4]byte{ip[0], ip[1], ip[2], ip[3]}) }
+
+// Parse an address (eg, "10.13.12.11") and return the corresponding IPv4 (eg, "10.13.12.11")
+func addrToIPv4(addr string) IPv4 {
+	ip := net.ParseIP(addr)
+	if ip == nil {
+		return IPv4{}
+	}
+	return ipToIPv4(ip)
+}
+
+// Parse a reverse address (eg, "11.12.13.10.in-addr.arpa.") and return the corresponding IPv4 (eg, "10.13.12.11")
+func raddrToIPv4(addr string) IPv4 {
+	ipStr := addr[:len(addr)-rdnsDomainLen]
```

Comment:
I think if you're going to have a standalone library function that returns a `ParseError` then it should be rigorous - just needs
````
    if !HasSuffix(addr, RDNSDomain) {
        return nil, &ParseError{"Invalid domain suffix", addr}
    }
````
at the top...